IdentityCore/src/cache/ios/MSIDKeychainTokenCache.m

```diff
@@ -22,30 +22,218 @@
+    
+    if (defaultKeychainGroup == s_defaultKeychainGroup)
+    {
+        return;
+    }
+    
+    if (!defaultKeychainGroup)
+    {
+        defaultKeychainGroup = [[NSBundle mainBundle] bundleIdentifier];
+    }
+    
+    s_defaultKeychainGroup = [defaultKeychainGroup copy];
+}
+
++ (MSIDKeychainTokenCache *)defaultKeychainCache
+{
+    static dispatch_once_t s_once;
+    
+    dispatch_once(&s_once, ^{
+        s_defaultCache = [[MSIDKeychainTokenCache alloc] init];
+    });
+    
+    return s_defaultCache;
+}
+
+- (nonnull instancetype)init
+{
+    return [self initWithGroup:s_defaultKeychainGroup];
+}
+
+- (nullable instancetype)initWithGroup:(nullable NSString *)keychainGroup
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    if (!keychainGroup)
+    {
+        keychainGroup = [[NSBundle mainBundle] bundleIdentifier];
+    }
+    
+    _keychainGroup = [MSIDKeychainUtil accessGroup:keychainGroup];
+    
+    if (!_keychainGroup)
+    {
+        return nil;
+    }
+    
+    MSID_LOG_INFO(nil, @"Using keychainGroup: %@", _PII_NULLIFY(_keychainGroup));
+    MSID_LOG_INFO_PII(nil, @"Using keychainGroup: %@", _keychainGroup);
+    
+    return self;
+}
+
+#pragma mark - MSIDTokenCacheDataSource
+
+- (BOOL)setItem:(MSIDToken *)item
+        withKey:(MSIDTokenCacheKey *)key
+     serializer:(id<MSIDTokenSerializer>)serializer
+        context:(id<MSIDRequestContext>)context
+          error:(NSError **)error
+{
+    assert(key);
+    assert(serializer);
+    
+    MSID_LOG_INFO(context, @"Set keychain item, key info (account: %@ service: %@)", _PII_NULLIFY(key.account), _PII_NULLIFY(key.service));
+    MSID_LOG_INFO_PII(context, @"Set keychain item, key info (account: %@ service: %@)", key.account, key.service);
+    MSID_LOG_INFO_PII(context, @"Item info %@", item);
+    
+    if (!key.service || !key.account)
+    {
+        if (error)
+        {
+            *error = [[NSError alloc] initWithDomain:MSIDErrorDomain code:MSIDErrorInternal userInfo:@{NSLocalizedDescriptionKey:@"Key is not valid. Make sure service and account are not nil."}];
+        }
+        MSID_LOG_ERROR(context, @"Set keychain item with invalid key.");
+        return NO;
+    }
+
+    NSData *itemData = [serializer serialize:item];
+    if (!itemData)
+    {
+        if (error)
+        {
+            *error = [[NSError alloc] initWithDomain:MSIDErrorDomain code:MSIDErrorInternal userInfo:@{NSLocalizedDescriptionKey:@"Failed to serialize token item."}];
+        }
+        MSID_LOG_ERROR(context, @"Failed to serialize token item.");
+        return NO;
+    }
+    
+    NSMutableDictionary *query = [self defaultQuery];
+    [query setObject:key.service forKey:(id)kSecAttrService];
+    [query setObject:key.account forKey:(id)kSecAttrAccount];
+    // Backward compatibility with ADAL.
+    [query setObject:[s_libraryString dataUsingEncoding:NSUTF8StringEncoding] forKey:(id)kSecAttrGeneric];
```

Comment:
can we add this (library string) to the default query?

Reply:
We need it only in "set" method. We won't be able to read MSAL cache if we add it to default query (default query is used in "get" methods)